### PR TITLE
SONARJS-887 Rule: super() must be invoked appropriately

### PR DESCRIPTION
**Pull request by ivandalbosco**

### CODE REVIEW

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+      int line = superTree.getFirstToken().line();
+      int column = superTree.getFirstToken().column();
+      MethodDeclarationTree method = getEnclosingConstructor(superTree);
+      Set<IssueLocation> secondaryLocations = new HashSet<>();
+
+      // get the usages of "super" before super()
+      Set<SuperTreeImpl> superTrees = new SuperDetector().detectIn(method);
+      superTrees.stream()
+        .filter(s -> !s.getParent().is(Kind.CALL_EXPRESSION))
+        .filter(s -> isBefore(s.getFirstToken(), line, column))
```
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java]: @ivandalbosco This is certainly a corner case but this approach may lead to FPs, e.g.:
```
class B51 extends A51 {
  constructor() {
    for (let i = 0; i < 1; super.x = 42) {
      super();
      i++;
    }
  }
}
```

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+    }
+  }
+
+  private void checkSuperHasCorrectNumberOfArguments(SuperTreeImpl superTree) {
+    if (isInConstructor(superTree) && !isInBaseClass(getEnclosingConstructor(superTree))) {
+      ClassTree classTree = getEnclosingClass(superTree);
+      ExpressionTree superClassTree = classTree.superClass();
+      // we consider only the simple case "class A extends B", not cases such as "class A extends class {} ..."
+      if (superClassTree.is(Kind.IDENTIFIER_REFERENCE)) {
+        IdentifierTree superClassId = (IdentifierTree) superClassTree;
```
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java]: @ivandalbosco The 3 lines below can be replaced with `Symbol superClassSymbol = superClassId.symbol();`

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+      }
+    }
+    return false;
+  }
+
+  private static ClassTree getEnclosingClass(Tree tree) {
+    return (ClassTree) CheckUtils.getFirstAncestor(tree, Kind.CLASS_DECLARATION, Kind.CLASS_EXPRESSION);
+  }
+
+  private void raiseIssue(SuperTreeImpl superTree, SuperTreeImpl firstSuperTree) {
```
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java]: @ivandalbosco Why do we have a separate method for that? It's invoked only once and its name does not really say what it does.

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3854.html [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  }
+}
+
+class FilaBrasileiro extends Dog {
+  constructor(name) {
+    super(name);
+  }
+}
+</pre>
+<p>&lt;h2&gt;Known Limitations&lt;/h2&gt;</p>
```
Comment (pynicolas) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3854.html]: @ivandalbosco That does not render correctly.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+  private void compareNumberOfArguments(SuperTreeImpl superTree, Symbol superClassSymbol) {
+    Optional<ClassTree> baseClassTree = getDeclarationTree(superClassSymbol);
+    if (baseClassTree.isPresent()) {
+      Optional<MethodDeclarationTree> baseClassConstructor = getConstructor(baseClassTree.get());
+      if (baseClassConstructor.isPresent()) {
+        Integer nbParams = baseClassConstructor.get().parameterList().size();
+        int nbArguments = ((CallExpressionTree) superTree.getParent()).arguments().parameters().size();
+        if (nbArguments != nbParams) {
+          String message = String.format(MESSAGE_SUPER_WITH_CORRECT_NUMBER_OF_ARGUMENTS, nbParams, nbParams == 1 ? "" : "s");
+          addIssue(superTree, message);
```
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java]: @ivandalbosco We could add a secondary location on the signature of the constructor of the super class.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+  @Override
+  public void visitFunctionExpression(FunctionExpressionTree tree) {
+    superInvocations.push(new LinkedList<>());
+    super.visitFunctionExpression(tree);
+    superInvocations.pop();
+  }
+
+  private void checkSuperOnlyInvokedInDerivedClassConstructor(SuperTreeImpl superTree) {
+    if (!isInConstructor(superTree) || isInBaseClass(getEnclosingConstructor(superTree))) {
```
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java]: @pynicolas If my understanding is correct, you'd like me to move up the test on isConstructor() and make it cover the 3 invocations of checkSuperXXX() methods in visitSuper(). But this does not work, as in one case (checkSuperOnlyInvokedInDerivedClassConstructor), the test is negated (with the "!" operator).

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+      addIssue(superTree, MESSAGE_SUPER_ONLY_IN_DERIVED_CLASS_CONSTRUCTOR);
+    }
+  }
+
+  private void checkSuperInvokedInAnyDerivedClassConstructor(MethodDeclarationTree method) {
+    if (isConstructor(method) && !isInBaseClass(method) && !isInDummyDerivedClass(method)) {
+      Set<SuperTreeImpl> superTrees = new SuperDetector().detectIn(method);
+      if (!superTrees.stream().anyMatch(s -> s.getParent().is(Kind.CALL_EXPRESSION))) {
```
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java]: @pynicolas You're right, thanks.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java (hunk 6 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java]: @pynicolas OK, done,

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java (hunk 1 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java]: @pynicolas I think I will live with that FP threat above my head. I assume that a constructor is expected to have a simple implementation, where super() is invoked almost immediately.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java (hunk 2 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java]: @pynicolas Great, thanks.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+  /**
+   * Returns true if the class of the specified method extends "null", else returns false.
+   * It is assumed that the class is a derived class.
+   */
+  private static boolean isInDummyDerivedClass(MethodDeclarationTree method) {
+    ClassTree classTree = getEnclosingClass(method);
+    return classTree.superClass().is(Kind.NULL_LITERAL);
+  }
+
+  private static MethodDeclarationTree getEnclosingConstructor(Tree tree) {
```
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java]: @pynicolas OK

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+    return null;
+  }
+
+  private static boolean isConstructor(FunctionTree tree) {
+    if (tree.is(Kind.METHOD)) {
+      MethodDeclarationTree constructor = (MethodDeclarationTree) tree;
+      Tree nameTree = constructor.name();
+      if (nameTree.is(Kind.IDENTIFIER_NAME)) {
+        String name = ((IdentifierTree) nameTree).name();
+        if ("constructor".equals(name)) {
```
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java]: @pynicolas OK

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java (hunk 3 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java]: @pynicolas You're right. Done.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/utils/CheckUtils.java [hunk 10]
```diff
@@ -106,5 +106,17 @@ public static boolean isDescendant(Tree tree, Tree potentialParent) {
     }
     return false;
   }
+
+  public static Tree getFirstAncestor(Tree tree, Kind... kind) {
```
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/utils/CheckUtils.java]: @pynicolas Well, I don't know. Annotations Nullable, CheckForNull, etc, are currently such a mess that they almost make more noise than help. They are in JSR 305, not in Java8. Looks more like a Findbugs stuff that like a Java spec. The Javadoc contract itself is hard to find on Google. I suggest we first decide what our group's policy is.

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3854.html (hunk 4 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3854.html]: @pynicolas Thanks, fixed.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java (hunk 5 above)
Comment (ivandalbosco) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/SuperInvocationCheck.java]: @pynicolas OK. I didn't find how to get the signature, so I defined the secondary location on the parameter list of the super constructor.

**Pull request merged**